### ISSUE TITLE
Fatal error:  Call to undefined method TGM_Plugin_Activation::get_instance()

### ISSUE THREAD

**Issue by drosa88** [TGMPA/TGM-Plugin-Activation#391]
Hi :-)

On Wordpress Multisites i get the following fatal error when i activate my plugin:

    Call to undefined method TGM_Plugin_Activation::get_instance()

I´m using this to display the notice also if the user closed it:

    TGM_Plugin_Activation::get_instance()->update_dismiss();

On basic wordpress installations works...do you have any idea? :-)

Greetings!

**Comment by jrfnl** [comment]
Which version of TGMPA are you using ? and can you show me [a backtrace of the error](https://gist.github.com/jrfnl/5925642) please ?

Also: TGMPA is currently not (fully) compatible with multi-site. This is on the road-map for v3.0.

And for TGMPA v2.5 (what's currently in develop), you ought to change your code to the below to avoid conflicts with themes/other plugins using TGMPA:
```php
$GLOBALS['tgmpa']->update_dismiss();
```

Or to be compatible with PHP 5.2 too (presuming you are in a local function scope) to:
```php
$tgmpa = call_user_func( array( get_class( $GLOBALS['tgmpa'] ), 'get_instance' ) );
$tgmpa->update_dismiss();
```

**Comment by drosa88** [comment]
Hi, thanks for your reply!

The backtrace error with my first used method call:
![bildschirmfoto 2015-05-06 um 15 24 03](https://cloud.githubusercontent.com/assets/6883050/7493840/554e3c46-f404-11e4-9747-b892f8df5e8e.png)

Error with your first code:
```
 Backtrace from warning 'Undefined index: tgmpa' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26:

    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 151 calling include()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 153 calling plugin_sandbox_scrape()


Fatal error: Call to a member function update_dismiss() on a non-object in /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php on line 26
```

Error with your second code (with compability to PHP 5.2 too):

```
 Backtrace from warning 'Undefined index: tgmpa' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26:

    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 151 calling include()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 153 calling plugin_sandbox_scrape()

Backtrace from warning 'get_class() called without object from outside a class' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26:

    /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26 calling get_class()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 151 calling include()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 153 calling plugin_sandbox_scrape()

Backtrace from warning 'call_user_func() expects parameter 1 to be a valid callback, first array member is not a valid class name or object' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26:

    /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php 26 calling call_user_func()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 151 calling include()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/plugins.php 153 calling plugin_sandbox_scrape()


Fatal error: Call to a member function update_dismiss() on a non-object in /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/better-facebook-page-box/better-facebook-page-box.php on line 27
```

Thanks for your support!

**Comment by jrfnl** [comment]
Ok, the error with the code I gave you is due to you being on v2.4.2, not on the newer (not yet released) version.

You could wrap the code in an `if` which checks for the version to make it compatible with both, but be aware that this will probably need to be adjusted again for v3.0 when it comes out:
```php
if ( defined( 'TGMPA_VERSION' ) ) {
	// new code
}
else {
	// old code
}
```

As for the rest of it, it sounds like you're trying to run your call too early, try hooking it in on `init` with prio 15.

**Comment by drosa88** [comment]
Thanks, when i load it in init-hook i receive no errors after activate the plugin.. but i don´t get the notification, that another plugin is required.. when i go to Design -> Install Plugins page i get the following:

```
 Backtrace from warning 'Undefined variable: newarray' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/wp-ulike/admin/classes/class-charts.php 113:

    /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/wp-ulike/admin/classes/class-charts.php 48 calling activities_dataset()
    calling enqueue_script()
    /homepages/9/d348610007/htdocs/foodarea/wp-includes/plugin.php 496 calling call_user_func_array()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/admin-header.php 80 calling do_action()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/admin.php 204 calling require_once()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/themes.php 10 calling require_once()

Backtrace from warning 'Undefined variable: newarray' at /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/wp-ulike/admin/classes/class-charts.php 113:

    /homepages/9/d348610007/htdocs/foodarea/wp-content/plugins/wp-ulike/admin/classes/class-charts.php 51 calling activities_dataset()
    calling enqueue_script()
    /homepages/9/d348610007/htdocs/foodarea/wp-includes/plugin.php 496 calling call_user_func_array()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/admin-header.php 80 calling do_action()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/admin.php 204 calling require_once()
    /homepages/9/d348610007/htdocs/foodarea/wp-admin/themes.php 10 calling require_once()
```

**Comment by jrfnl** [comment]
The error as you say, does not come from the TGMPA class.

Regarding the notice - I kind of think this might already be solved in `develop`, so I suggest you try that first.

**Comment by drosa88** [comment]
In included your dev-version now. After activate i get the following:

```
Fatal error: Call to protected TGM_Plugin_Activation::__construct() from invalid context in /homepages/9/d348610007/htdocs/foodarea/wp-content/themes/blogary/functions/class/class-tgm-plugin-activation.php on line 953
```

Maybe the problem is that my theme is using your class too?

**Comment by jrfnl** [comment]
backtrace ?

And it should be fine to use TGMPA both in a theme as well as plugin. That's how it's designed. However, TGMPA should be called properly using the code from the `example.php` file. Implementations which divert from that code , i.e. try to call `__construct()` directly not using the `tgmpa()` function are impossible for us to support.

**Comment by drosa88** [comment]
Yes, backtrace..
![bildschirmfoto 2015-05-06 um 16 54 33](https://cloud.githubusercontent.com/assets/6883050/7495822/9e3b604e-f410-11e4-89fa-4c338e2ae402.png)

So it´s an issue from the theme developer itself...not my problem as the developer of the plugin.. :D

**Comment by jrfnl** [comment]
Though you might want to inform the theme developer....

**Comment by drosa88** [comment]
Sure! Thanks for your great support!

**Comment by jrfnl** [comment]
@drosa88 I just checked through some old versions - it could be that the theme includes a really old version of TGMPA (2.3.6 or less - over two years old) which causes that error message. Please urge them to upgrade the TGMPA version they include.

**Comment by drosa88** [comment]
Thanks for your information, i wrote him an email! :)

**Comment by GaryJones** [comment]
Can this be closed then since it's not about the recent release or develop branch?

**Comment by drosa88** [comment]
Yes, on other installations is working great. thanks! :)

**Comment by jrfnl** [comment]
@GaryJones We may want to add this to an FAQ somewhere - don't believe we have one at this moment though.

**Comment by zackkatz** [comment]
@jrfnl @GaryJones Just had a client white screen of death after upgrading; the Avada theme uses class constructor instead of function: 

```
Fatal error: Call to protected TGM_Plugin_Activation::__construct() from invalid context in
[...]/httpdocs/wp-content/themes/Avada/framework/class-tgm-plugin-activation.php
on line 1006
```

Obviously, the Avada theme is installed by a lot of people. Maybe not making constructor public with a notice would be better. Thoughts?

**Comment by GaryJones** [comment]
> Thoughts?

I think Avada author(s) should follow our usage instructions ;-)

If / When they do a new Avada release to update to TGMPA 2.5.0, they can update their initialising code too. It's not something I think we need to worry about right now.

**Comment by jrfnl** [comment]
@zackkatz see my answer a few messages above: https://github.com/TGMPA/TGM-Plugin-Activation/issues/391#issuecomment-99614994
I suggest the user contacts Avada and urges them to upgrade.

**Comment by zackkatz** [comment]
@GaryJones @jrfnl Sorry for not RTFC. I completely agree they should have already updated the code and followed instructions. 

The Avada dev(s) haven't done that though, and are one of the bigger WP themes. 137,070 copies sold…maybe there should be a exception made just for them.

The alternative for me is downgrading to the previous version, which I suppose I should do.

**Comment by GaryJones** [comment]
I'm loathe to make exceptions for one product which hasn't "played by the rules", at the expense of those hundreds or thousands of products which have.

The power to fix this is in the hands of the Avada developers.

**Comment by jrfnl** [comment]
@zackkatz The code snippet they are using is based on the two year old v2.3.6 and has been deprecated for over a year. On top of that v2.3.6 contains known security and extensibility issues. What more can I say ?

**Comment by zackkatz** [comment]
@jrfnl @GaryJones I totally understand. Avada theme has been nothing but headaches for a few of my plugins. Thanks for your work on this library :+1:

**Comment by jrfnl** [comment]
@zackkatz FYI - even though we still advocate themes and plugins to upgrade. The scale at which v2.3.6 and lower version are still being used turns out to be daunting. So we've decided to revert the compatibility break - see PR #448. I expect it will be released as v 2.5.2 today or tomorrow.
